stdlib/public/Platform/musl.modulemap

```diff
@@ -0,0 +1,20 @@
+//===--- musl.modulemap ---------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+module SwiftMusl [system] {
+  header "SwiftMusl.h"
```

Comment:
It should be possible to inject this modulemap file into the Musl directory using LLVM VFS. 
We already use this mechanism to inject `glibc.modulemap` into the Glibc installation (`swift::getClangInvocationFileMapping`). That way the Musl headers could be referenced directly from the modulemap instead of using an additional `SwiftMusl.h` header, preventing another module from hijacking the system headers: with the current approach another C module that includes the same system header might take ownership over the header, which would cause compile errors in some cases.

Reply:
Any plans to remove the `SwiftGlibc.h` header soon?

Reply:
I'm a bit confused here, is the approach suggested in https://forums.swift.org/t/problems-with-swiftglibc-and-proposed-fix/37594 no longer applicable? Wasn't a single header referenced from the modulemap, importing the rest of the headers the preferred solution?

Reply:
That approach has a flaw that might cause a compile error related to the modularization in the following scenario:

Consider this modulemap:
```
module B {
  header "b.h"
}
module C {
  header "c.h"
}
```
and three headers:
```cpp
// a.h
class Decl {}
```
```cpp
// b.h
#include "a.h"
```
```cpp
// c.h
#include "c.h"
```

Notice that `a.h` is not referenced from the modulemap. As the forum post mentions, this is not an issue for Clang:

> If the platform does not provide a module map for a given header, it will be included textually, both into the SwiftGlibc module and into any imported C / Objective-C / C++ modules that include the header. All of these modules will therefore contain the definitions from this header.
> This is fine under Clang’s rules for modules; Clang has logic to deduplicate definitions caused by textual inclusion of the same header into multiple modules.

However, an important part that the forum post misses is that while _Clang_ can handle this, _Swift_ cannot. What Swift does in this case is it assumes that the first header to include `a.h` owns the header, in a sense that `a.h` is assigned to the same module as either `b.h` or `c.h`, but we can't really say which one is it.

This situation might occur in the system header scenario: assume we're compiling a Swift program that imports two C modules: `SwiftGlibc`/`SwiftMusl` and some other user-provided module. If the latter also contains an `#include <some_stdlib_header.h>`, Swift might treat `some_stdlib_header.h` as a part of the user-provided module. This will cause a compile error for a perfectly valid code:
```swift
import SwiftMusl
someFunctionFromTheStdlib()
// error: user-provided module not imported
```

The injection approach avoids [both of the issues described in the forum post](https://forums.swift.org/t/problems-with-swiftglibc-and-proposed-fix/37594#summary-1): symbols such as `pid_t` are available, and there is no cyclic dependency between libc++ and SwiftGlibc. I can explain this in a bit more detail if you'd like.

At the moment [libstdc++ modulemap](https://github.com/apple/swift/blob/51c5da0f705fe3187cdd681331bdc6c835e0faa0/stdlib/public/Cxx/std/libstdcxx.modulemap) relies on the new injection approach, while SwiftGlibc still has a `SwiftGlibc.h` header with a list of `#include<>`-s. The idea is to eventually get rid of `SwiftGlibc.h` and refer to all of the glibc headers from the modulemap directly.

> Any plans to remove the SwiftGlibc.h header soon?

@buttaface no specific timeline as of now, but contributions are welcome 🙂

Reply:
Just to clarify, I don't know if it's strictly necessary to use the injection approach for SwiftMusl right now, I'm just pointing out an issue that might arise in the future.

Reply:
@MaxDesiatov, I brought this up so that you could try out this new approach with Musl too, if interested.

Reply:
> It should be possible to inject this modulemap file into the Musl directory using LLVM VFS.
> We already use this mechanism to inject `glibc.modulemap` into the Glibc installation (`swift::getClangInvocationFileMapping`).

Slightly off-topic here, but does this mechanism work on Windows? Looks like it could solve a big pain on Windows installation. Nowadays Swift on Windows installer will copy the `modulemap`s to UCRT/WinSDK/MSVC accordingly. This is somehow fragile because we need to repeat this process every time after the environment is updated, and since the detection is hard-coded we cannot support any version of Visual Studio at the time.

Reply:
(On the Windows front, I believe this is actually what the Windows build does now.)